sycl/doc/extensions/experimental/sycl_ext_oneapi_complex.asciidoc

```diff
@@ -42,17 +42,12 @@ SYCL specification refer to that revision.
 
 == Status
 
-This is a proposed extension specification, intended to gather community
-feedback.  Interfaces defined in this specification may not be implemented yet
-or may be in a preliminary state.  The specification itself may also change in
-incompatible ways before it is finalized.  *Shipping software products should
-not rely on APIs defined in this specification.*
-
-[NOTE]
-====
-This extension is not currently implemented in {dpcpp}.
-====
-
+This is an experimental extension specification, intended to provide early
+access to features and gather community feedback. Interfaces defined in this
+specification are implemented in {dpcpp}, but they are not finalized and may
+change incompatibly in future versions of {dpcpp} without prior notice.
+*Shipping software products should not rely on APIs defined in this
+specification.*
```

Comment:
Github won't let me comment on lines that didn't change, so here are some global comments:

* The namespace must be changed to `sycl::ext::oneapi::experimental`.  All experimental extensions must be in the `experimental` namespace.

* You might consider changing the "Feature test macro" section to use the wording in the template "The APIs of this experimental extension are not versioned, so the feature-test macro always has this value".  Versioning for the feature-test macro doesn't really make sense if you plan to make API breaking changes.

* The member function "complex" operators allow the RHS to be a different element type:

  ```
  template<typename X>
  complex& operator+=(const complex<X>&);
  ```

  Whereas the free functions require the LHS and RHS to have the same element type:

  ```
  template<typename T>
  complex<T> operator+(const complex<T>&, const complex<T>&);
  ```

  Why is that?  Is it a standard C++ convention?

* I think you have a typo in the synopsis here:

  ```
  // operator=
  template<typename X>
  complex& operator+=(const complex<X>&);
  ```

  It should be `operator=` (not `operator+=`).  You are also missing this operator from the table describing all the operators.

* This statement seems a little weird:

  > The class `sycl::oneapi::complex` class, has specializations of `T`; `float`, `double`, and `sycl::half` defined.

  I assume we mean that the application may only uses these types for the `T` template parameter?  We normally express this sort of thing in the SYCL spec by saying:

  > The `T` template parameter must be one of the types `float`, `double`, or `sycl::half`.

  (And I would remove the synopsis showing the three specializations.)

* In the table describing all the operators, please remove all uses of `gencomplex` and `genfloat`.  Instead, show these operators using the template syntax you have in the synopsis.  We've had a lot of confusion in the SYCL spec about the meaning of these "gen" types, but the recent consensus is that these are a shorthand for describing a bunch of overloads.  Since you want to use template functions (and not overloads), it is inappropriate to use these "gen" types here.  (See KhronosGroup/SYCL-Docs#428 if you want more on this.)

* I assume the functions specified in "Mathematical operations" are supposed to be a bunch of overloads.  I think the clearest way to describe this is to just write them all out.  There are only three possibilities for `gencomplex`, so it's not worth the confusion to use these "gen" types.  For example `abs` would be specified like:

  ```
  float abs(const complex<float>& x)
  double abs(const complex<double>& x)
  sycl::half abs(const complex<sycl::half>& x)
  ```

  This also makes it crystal clear what the return type is.  Before you had to guess which `genfloat` went with each `gencomplex`.

Reply:
Thanks for your review, @gmlueck.

The following commit should resolve all your points:

- First, since this proposal has been merged, some changes have been made to the implementations; The synopsis of the complex class is updated to reflect that and mirror the open Khronos proposal.

- Namespace is changed to ```sycl::ext::oneapi::experimental```

- Feature test macro uses the template's wording for experimental extension

- The binary operators +=, -= ... allow the ```rhs``` argument to be of complex type with the value_type different from the ```lhs``` complex. Whereas the binary operator +, - ... does not allow that but allows for the ```rhs``` or ```lhs``` to be a scalar type. Types that can be used to construct a complex.
This behaviour has been inherited from ```std::complex```, since the complex extension is based on.

- Fix the typo of the operator, and simplify the extension to have the description within the synopsis (which removes the possibility of creating/adding typo).

- Use your wording for explaining the template parameter ```T```

- With the simplification mentioned previously, the tables do not exist anymore, and I propose to have the mathematical functions declared using template <typename T> instead of ```gencomplex``` and ```genfloat```. Do you think that's clear enough?